### PR TITLE
Project selection dialog

### PR DESCRIPTION
**Pull request by ivanporty**
Final part of project selection change - dialog.
This looks like a large PR but it's actually not - at least 400 LOC are form XML/copyright headers/imports/visible for testing lines, and the rest cannot be split logically and should be straightforward to review.

You can review FilteredTextTableCellRenderer.java and ProjectListTableModelTest.java and their tests first - those are very small and independent, and continue with ProjectSelectionDialog.

Initial state - no logged in users:
sign_in_screen

Logged in users, no project selection:
dialog

Filtered projects - both in our field and IDEA speed search on table (they are synced):
dialog-filter

### CODE REVIEW

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectListTableModel.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  @Override
+  public Object getValueAt(int row, int column) {
+    switch (column) {
+      case PROJECT_NAME_COLUMN:
+        return projectList.get(row).getName();
+      case PROJECT_ID_COLUMN:
+        return projectList.get(row).getProjectId();
+    }
+
+    return "";
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectListTableModel.java]: if you are using a switch, why not just use a `default` case?

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectListTableModel.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  @Override
+  public String getColumnName(int column) {
+    switch (column) {
+      case PROJECT_NAME_COLUMN:
+        return GctBundle.getString("project.selector.project.list.project.name.column");
+      case PROJECT_ID_COLUMN:
+        return GctBundle.getString("project.selector.project.list.project.id.column");
+    }
+
+    return "";
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectListTableModel.java]: same comment

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectorSignInPanel.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import java.awt.event.ActionEvent;
+import javax.swing.AbstractAction;
+import javax.swing.Action;
+import javax.swing.JButton;
+
+/** Sign in panel shown in {@link ProjectSelectionDialog} when no users are logged in. */
+class ProjectSelectorSignInPanel extends BaseGoogleLoginUi {
+
+  ProjectSelectorSignInPanel() {
+    super(GctBundle.message("select.project.signin"));
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectorSignInPanel.java]: Didn't catch this last time time, but perhaps a better name here? Something along the lines of `cloud.project.selector.signin.message`

**Comment by etanshaul** [comment]
Instead of "Filter Name/ID", how about "Search by name or ID"?

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectorSignInPanel.java (hunk 3 above)
Comment (ivanporty) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectorSignInPanel.java]: good idea, changing here and adding cloud.project.selector to other strings instead of just project.selector

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectListTableModel.java (hunk 1 above)
Comment (ivanporty) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectListTableModel.java]: done.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectListTableModel.java (hunk 2 above)
Comment (ivanporty) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectListTableModel.java]: done.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java [hunk 4]
```diff
@@ -33,6 +95,334 @@
...
+
+  @VisibleForTesting
+  CloudProject getCloudProject() {
+    CredentialedUser user = accountComboBox.getItemAt(accountComboBox.getSelectedIndex());
+    return CloudProject.create(getSelectedProjectName(), user.getEmail());
+  }
+
+  @VisibleForTesting
+  void loadUsersAndProjects() {
+    Collection<CredentialedUser> credentialedUsers =
```
[the remark this answers is not in the document]
Comment (ivanporty) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: It does select active user by default, see line 152, but will keep the order of users in the dropdown as it is. I can change it to have the active user on top.

[the remark this answers is not in the document]
Comment (etanshaul) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: oops, didn't catch that. I think its fine as is.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java [hunk 5]
```diff
@@ -33,6 +95,334 @@
...
+
+    DialogManager.show(dialogWrapper);
+
+    int result = dialogWrapper.getExitCode();
+    return result == DialogWrapper.OK_EXIT_CODE ? getCloudProject() : null;
+  }
+
+  @VisibleForTesting
+  void setCloudProject(CloudProject cloudProject) {
+    this.cloudProject = cloudProject;
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: this also feels like an opportunity for refactoring. setting this as a member variable feels off to me.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java [hunk 6]
```diff
@@ -33,6 +95,334 @@
...
+    updateProjectAccountInformation();
```
[the remark this answers is not in the document]
Comment (ivanporty) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: In loadUsersAndProjects it loads all accounts and their projects and selects active (default) account - i.e. prepares default state of the dialog.
In updateProjectAccountInformation it changes the state of the dialog to match that of the passed CloudProject. How about setProjectSelectionTo(CloudProject)?

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java (hunk 5 above)
Comment (ivanporty) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: We can remove this variable and just use it once to set the state of the account combobox and project list selection - I think we won't need to keep it after that. WDYT?

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java [hunk 7]
```diff
@@ -33,6 +95,334 @@
    */
   @Nullable
   CloudProject showDialog(@Nullable CloudProject cloudProject) {
-    return cloudProject;
+    // create dialog wrapper for the form once, create services and subscribe to login changes once.
+    if (dialogWrapper == null) {
+      dialogWrapper = new ProjectSelectionDialogWrapper();
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: I'm trying to wrap my head around this. Why didn't you want this class to be a `DialogWrapper` itself?

Comment (etanshaul) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: in other words, why not just extend from DialogWrapper? As is it is, it is very tightly coupled to DialogWrapper anyway, so I'm not seeing the advantage. I could be wrong though.

Comment (ivanporty) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: I tried it first and inheriting from `DialogWrapper` brings a lot of internal UI implementation with it - I had to spy/mock/override a lot of internal methods, and it started 'leaking projects' because I used it - something was not disposing properly in mock test mode. It's much easier and cleaner to test and edit as a simple form, but we have to have this wrapper class to pass dialog wrapper buttons and additional actions. It is not very friendly to composition instead of inheritance.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (ivanporty) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: Removed updateProjectAccountInformation() altogether.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java (hunk 5 above)
Comment (ivanporty) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: done.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java [hunk 8]
```diff
@@ -33,6 +103,346 @@
...
+
+  @VisibleForTesting
+  void setCloudProject(CloudProject cloudProject) {
+    if (cloudProject != null && !Strings.isNullOrEmpty(cloudProject.googleUsername())) {
+      Optional<CredentialedUser> loggedInUser =
+          Services.getLoginService().getLoggedInUser(cloudProject.googleUsername());
+      if (loggedInUser.isPresent()) {
+        selectedProjectsByAccount.put(loggedInUser.get(), cloudProject.projectName());
+        accountComboBox.setSelectedItem(loggedInUser.get());
+        updateProjectList(loggedInUser.get());
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: why is this needed here if the project list is already loaded (and cached) in #loadUsersAndProjects ?

Comment (etanshaul) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: oh..I think i see..thinking about this

Comment (etanshaul) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: actually, nm. If the cached list is already loaded, why do we need to load it again for this specific user?

Comment (ivanporty) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: It won't be loaded since it's cached, I sort of added it for increased readability so reader doesn't need to imply project will be selected because of the combox box event. We can remove it safely.

Comment (etanshaul) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: ah ok. lets remove it then. it had the opposite effect on me. made me confused. :)

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java [hunk 9]
```diff
@@ -33,6 +103,346 @@
...
+              project ->
+                  project
+                      .getMessageBus()
+                      .connect(dialogWrapper.getDisposable() /* disconnect once dialog is gone. */)
+                      .subscribe(
+                          GoogleLoginListener.GOOGLE_LOGIN_LISTENER_TOPIC,
+                          this::loadUsersAndProjects));
+    }
+
+    loadUsersAndProjects();
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: I know we can go back and forth forever on naming, but just thinking after looking again..how about:

loadUsersAndProjects -> loadAllProjects
setCloudProject -> setActiveProject (I think the "cloud" part is implied enough)

Comment (etanshaul) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: actually not setActiveProject, but setSelectedProject

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java (hunk 8 above)
Comment (ivanporty) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: done.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java (hunk 9 above)
Comment (ivanporty) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: sounds good, and I'll change the getter name to reflect that idea as well - i.e. `getSelectedProject`

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java [hunk 10]
```diff
@@ -33,6 +103,348 @@
...
+        Services.getLoginService().getAllUsers().values();
+    if (credentialedUsers.isEmpty()) {
+      showSignInRequest();
+    } else {
+      hideSignInRequest();
+      cachedProjectList.clear();
+      accountComboBox.removeAllItems();
+      for (CredentialedUser user : credentialedUsers) {
+        accountComboBox.addItem(user);
+        updateProjectList(user);
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: This is what I was thinking, lmkwyt

- `loadAllProjects` is responsible for loading and caching the projects for all users. Thats is; it doesnt update the UI.
- `setSelectedProject` then updates the dropdown with the current CloudProject, and in turn, causes the table model to update (pulling the list of projects from the cache).
- the listener on the account combobox updates the table model like in the above bullet

Its similar, but I think this will separate concerns a bit more cleanly. Sorry if I am contradicting previous things I've said with these comments. The motivation is I think that updateProjectList is a bit too complicated and doing too much (loading projects, updating the UI..).

Comment (ivanporty) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: sounds reasonable, I'll do a simple refactor of this idea, trying to clean up `updateProjectList' a bit more

**Comment by etanshaul** [comment]
Can be done in a separate PR later, but it might be nice to have some loading indicator when the projects are being refreshed (or when a new account is selected and the project list is loading).

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java [hunk 11]
```diff
@@ -16,15 +16,85 @@
...
+  private RefreshAction refreshAction;
+
+  private ProjectListTableModel projectListTableModel;
+
+  private ProjectLoader projectLoader = new ProjectLoader();
+
+  private Map<CredentialedUser, ListenableFuture<List<Project>>> runningProjectLoaderJobs =
+      Maps.newHashMap();
+  // last project name selection made by user for an account.
+  private Map<CredentialedUser, String> selectedProjectsByAccount = Maps.newHashMap();
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: what do you think about combining selecteProjectsByAccount and cachedProjectList into a single mutlimap?

Comment (etanshaul) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: specifically, guava's multimap

Comment (ivanporty) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: you mean selected project is marked somehow in project list, i.e. we have Multimap<User,Project> and selected project is one of them?

Comment (etanshaul) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: Never mind, it may have been a bad idea.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java [hunk 12]
```diff
@@ -33,6 +103,348 @@
...
+    return result == DialogWrapper.OK_EXIT_CODE ? getSelectedProject() : null;
+  }
+
+  @VisibleForTesting
+  void setSelectedProject(CloudProject cloudProject) {
+    if (cloudProject != null && !Strings.isNullOrEmpty(cloudProject.googleUsername())) {
+      Optional<CredentialedUser> loggedInUser =
+          Services.getLoginService().getLoggedInUser(cloudProject.googleUsername());
+      if (loggedInUser.isPresent()) {
+        selectedProjectsByAccount.put(loggedInUser.get(), cloudProject.projectName());
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: Cloud project name is not unique; project ID is. I just want to make sure the logic here and elsewhere is aware of this. In other words, if this is actually using the cloud project name here, then I think this example usage will have problems (since it may not select the correct project).

Comment (ivanporty) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: so in `AppEngineDeploymentConfiguration`, the field is `cloudProjectName`, is it actually name or ID?

Comment (etanshaul) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: The main thing is that whenever we are keying on it, we must use the ID. But if we are displaying it (like in the new Cloud Project widget - [name] / [email]) we use the name.

Comment (ivanporty) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: sounds good, then let's add  `projectId` to keep the actual ID in `CloudProject` too

**Comment by ivanporty** [comment]
Cleaned up `updateProjectList` a bit more, renamed to better understand what it does.
Also, added loading indicator - using small neat indicator in top right corner from IDEA table.
PTAL at changes.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java [hunk 13]
```diff
@@ -33,6 +104,363 @@
...
+      accountComboBox.removeAllItems();
+      for (CredentialedUser user : credentialedUsers) {
+        accountComboBox.addItem(user);
+        loadProjectList(user);
+      }
+    }
+  }
+
+  private void loadProjectList(CredentialedUser user) {
+    // check if cached, if not then check if already loading, and finally start loading if not.
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: could you promote top level comments like this that apply to the whole method to actually be javadoc method comments? I know some of these cases are for private methods, but I still think it makes sense.

**Comment by etanshaul** [comment]
I dunno if its a bug or not, but the loader doesn't consistently show up. I just clicked the refresh button a bunch of times and sometimes it appeared and sometimes not.

Also, maybe move the loader to be beside the refresh button? It took me several seconds to find it even when it did appear.

**Comment by etanshaul** [comment]
or even in the middle of table (if thats possible). wherever you think makes more sense.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java [hunk 14]
```diff
@@ -33,6 +104,363 @@
...
+  }
+
+  @VisibleForTesting
+  CloudProject getSelectedProject() {
+    CredentialedUser user = accountComboBox.getItemAt(accountComboBox.getSelectedIndex());
+    return CloudProject.create(getSelectedProjectName(), getSelectedProjectId(), user.getEmail());
+  }
+
+  @VisibleForTesting
+  void loadAllProjects() {
```
Comment (etanshaul) [inline google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: thanks for the refactoring. much easier to read now

**Comment by etanshaul** [comment]
possible bug:

1) sign in to multiple accounts
2) select the second account in the list
3) click refresh

-> I'd expect the same account's projects to be refreshed. Instead it sent me back to the other account.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java (hunk 13 above)
Comment (ivanporty) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: sure, great idea.

File: google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java (hunk 14 above)
Comment (ivanporty) [reply google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java]: yeah it is much easier to read, good catch!

**Comment by ivanporty** [comment]
Added clearly visible centered overlay progress bar, preserved account selection when refresh button is hit.
